# Cantera Tutorial: Python

## Getting Started

Open a new Jupyter Notebook. If unfamiliar with Jupyter Notebooks see the [Jupyter Documentation](https://jupyter.readthedocs.io/en/latest/install.html) for installation and basic use instructions. Import the Cantera Python module and NumPy by running:

In [1]:
import cantera as ct
import numpy as np

When using Cantera, the first thing you usually need is an object representing some phase of matter. Here, we'll create a gas mixture:

In [2]:
gas1 = ct.Solution('gri30.xml')

To view the state of the mixture, *call* the `gas1` object as if it were a function:

In [3]:
gas1()


  gri30:

       temperature             300  K
          pressure          101325  Pa
           density       0.0818891  kg/m^3
  mean mol. weight         2.01588  amu

                          1 kg            1 kmol
                       -----------      ------------
          enthalpy           26470        5.336e+04     J
   internal energy     -1.2109e+06       -2.441e+06     J
           entropy           64914        1.309e+05     J/K
    Gibbs function     -1.9448e+07        -3.92e+07     J
 heat capacity c_p           14312        2.885e+04     J/K
 heat capacity c_v           10187        2.054e+04     J/K

                           X                 Y          Chem. Pot. / RT
                     -------------     ------------     ------------
                H2              1                1         -15.7173
     [  +52 minor]              0                0



What you have just done is created an object `gas1` that implements GRI-Mech 3.0, the 53-species, 325-reaction natural gas combustion mechanism developed by Gregory P. Smith, David M. Golden, Michael Frenklach, Nigel W. Moriarty, Boris Eiteneer, Mikhail Goldenberg, C. Thomas Bowman, Ronald K. Hanson, Soonho Song, William C. Gardiner, Jr., Vitali V. Lissianski, and Zhiwei Qin.  See the [GRI-Mech Home Page](http://combustion.berkeley.edu/gri-mech/) for more information.

The `gas1` object has properties you would expect for a gas mixture: a temperature, a pressure, species mole and mass fractions, etc. As we will soon see, it has many more properties.

The summary of the state of `gas1` that you found above shows that the new objects created from the `gri30.xml` input file start out with a temperature of 300 K, a pressure of 1 atm, and have a composition that consists of only one species, in this case hydrogen. There is nothing special about H2 - it just happens to be the first species listed in the input file defining GRI-Mech 3.0. In general, whichever species is listed first will initially have a mole fraction of 1.0, and all others will be zero.

## Setting the State

The state of the object can easily be changed. For example:

In [4]:
gas1.TP = 1200, 101325

sets the temperature to 1200 K and the pressure to 101325 Pa (Cantera always uses SI Units). After this statement, calling `gas1()` results in:

In [5]:
gas1()


  gri30:

       temperature            1200  K
          pressure          101325  Pa
           density       0.0204723  kg/m^3
  mean mol. weight         2.01588  amu

                          1 kg            1 kmol
                       -----------      ------------
          enthalpy      1.3296e+07         2.68e+07     J
   internal energy      8.3462e+06        1.682e+07     J
           entropy           85228        1.718e+05     J/K
    Gibbs function     -8.8978e+07       -1.794e+08     J
 heat capacity c_p           15378          3.1e+04     J/K
 heat capacity c_v           11253        2.269e+04     J/K

                           X                 Y          Chem. Pot. / RT
                     -------------     ------------     ------------
                H2              1                1         -17.9775
     [  +52 minor]              0                0



Notice that the temperature has been changed as requested, but the density has changed too. The pressure and the composition have not.

Thermodynamics generally requires that *two* properties in addition to composition information be specified to fix the intensive state of a substance (or mixture). The state of the mixture can be set using several combinations of two properties. The following are all equivalent:

In [6]:
gas1.TP = 1200, 101325            # temperature, pressure
gas1.TD = 1200, 0.0204723         # temperature, density
gas1.HP = 1.32956e7, 101325       # specific enthalpy, pressure
gas1.UV = 8.34619e6, 1/0.0204723  # specific internal energy, specific volume
gas1.SP = 85227.6, 101325         # specific entropy, pressure
gas1.SV = 85227.6, 1/0.0204723    # specific entropy, specific volume

In each case, the values of the extensive properties must be entered *per unit mass*.

Properties may be read independently, such as

In [7]:
gas1.T

1200.0044548350836

or

In [8]:
gas1.h

13295636.190310445

or together:

In [9]:
gas1.UV

(8346238.627182945, 48.84649013545132)

The composition can be set in terms of either mole fractions (`X`) or mass fractions (`Y`):

In [10]:
gas1.X = 'CH4:1, O2:2, N2:7.52'

Mass and mole fractions can also be set using the `dict` object, for cases where the composition is stored in a variable or being computed:

In [11]:
phi = 0.8
gas1.X = {'CH4':1, 'O2':2/phi, 'N2': 2*3.76/phi}

When the composition alone is changed, the temperature and density are held constant. This means that the pressure and other intensive properties will change. The composition can also be set in conjunction with the intensive properties of the mixture:

In [12]:
gas1.TPX = 1200, 101325, 'CH4:1, O2:2, N2:7.52'
gas1()


  gri30:

       temperature            1200  K
          pressure          101325  Pa
           density        0.280629  kg/m^3
  mean mol. weight         27.6332  amu

                          1 kg            1 kmol
                       -----------      ------------
          enthalpy      8.6194e+05        2.382e+07     J
   internal energy      5.0088e+05        1.384e+07     J
           entropy          8914.3        2.463e+05     J/K
    Gibbs function     -9.8352e+06       -2.718e+08     J
 heat capacity c_p          1397.3        3.861e+04     J/K
 heat capacity c_v          1096.4         3.03e+04     J/K

                           X                 Y          Chem. Pot. / RT
                     -------------     ------------     ------------
                O2       0.190114         0.220149         -28.7472
               CH4       0.095057        0.0551863          -35.961
                N2       0.714829         0.724665         -25.6789
     [  +50 minor]        

The composition above was specified using a string. The format is a comma-separated list of `<species name>:<relative mole numbers>` pairs. The mole numbers will be normalized to produce the mole fractions, and therefore they are "relative" mole numbers. Mass fractions can be set this way too by changing `X` to `Y` in the above statements.

The composition can also be set using an array, which must have the same size as the number of species. For example, to set all 53 mole fractions to the same value, do this:

In [13]:
gas1.X = np.ones(53)  # NumPy array of 53 ones

Or, to set all the mass fractions to equal values:

In [14]:
gas1.Y = np.ones(53)

When setting the state, you can control what properties are held constant by passing the special value `None` to the property setter. For example, to change the specific volume to 2.1 m<sup>3</sup>/kg while holding entropy constant:

In [15]:
gas1.SV = None, 2.1

Or to set the mass fractions while holding temperature and pressure constant:

In [16]:
gas1.TPX = None, None, 'CH4:1.0, O2:0.5'

## Working with a Subset of Species

Many properties of a [`Solution`](https://cantera.org/documentation/dev/sphinx/html/cython/importing.html#cantera.Solution) provide values for each species present in the phase. If you want to get values only for a subset of these species, you can use Python's "slicing" syntax to select data for just the species of interest. To get the mole fractions of just the major species in `gas1`, in the order specified, you can write:

In [17]:
Xmajor = gas1['CH4','O2','CO2','H2O','N2'].X

If you want to use the same set of species repeatedly, you can keep a reference to the sliced phase object:

In [18]:
major = gas1['CH4','O2','CO2','H2O','N2']
cp_major = major.partial_molar_cp
wdot_major = major.net_production_rates

The slice object and the original object share the same internal state, so modifications to one will affect the other.

## Working with Mechanism Files

In the previous example, we created an object that models an ideal gas mixture with the species and reactions of GRI-Mech 3.0, using the `gri30.xml` input file included with Cantera. This is a "pre-processed" XML input file written in a format that is easy for Cantera to parse, but hard for humans to write. Cantera also supports an input file format that is easier for humans to write, called *CTI*. Several reaction mechanism files in this format are included with Cantera, including ones that model high-temperature air, a hydrogen/oxygen reaction mechanism, and a few surface reaction mechanisms. These files are usually located in the `data` subdirectory of the Cantera installation directory, e.g. `C:\\Program Files\\Cantera\\data` on Windows or `/usr/local/cantera/data/` on Unix/Linux/Mac OS X machines, depending on how you installed Cantera and the options you specified.

If, for some reason, Cantera has difficulty finding where these files are on your system, set environment variable `CANTERA_DATA` to the directory or directories (separated using `;` on Windows or `:` on other operating systems) where they are located. Alternatively, you can call function [`add_directory`](https://cantera.org/documentation/dev/sphinx/html/cython/importing.html#cantera.add_directory) to add a directory to the Cantera search path:

In [19]:
ct.add_directory('/usr/local/cantera/my_data_files')

Cantera input files are plain text files, and can be created with any text editor. See the guide *[Defining Phases](https://cantera.org/tutorials/cti/defining-phases.html)* for more information.

A Cantera input file may contain more than one phase specification, or may contain specifications of interfaces (surfaces). Here, we import definitions of two bulk phases and the interface between them from the file `diamond.cti`:

In [20]:
gas2 = ct.Solution('diamond.cti', 'gas')
diamond = ct.Solution('diamond.cti', 'diamond')
diamond_surf = ct.Interface('diamond.cti', 'diamond_100', [gas2, diamond])

Note that the bulk (i.e., 3D or homogenous) phases that participate in the surface reactions must also be passed as arguments to [`Interface`](https://cantera.org/documentation/dev/sphinx/html/cython/importing.html#cantera.Interface).

### Converting CK-format files

See *[Converting CK-format files](https://cantera.org/tutorials/ck2cti-tutorial.html)* in the *[Working with Input Files](https://cantera.org/tutorials/input-files.html)* documentation.

## Getting Help

In addition to the *[Python Module Documentation](https://cantera.org/documentation/index.html)*, documentation of the Python classes and their methods can be accessed from within the Python interpreter as well.

Suppose you have created a Cantera object and want to know what methads are avialable for it, and get help on using the methods:

In [21]:
g = ct.Solution('gri30.xml')

To get help on the Python class that this object is an instance of, put a question mark `?` after the variable:

In [22]:
g?

For a simple list of the properties and methods of this object:

In [23]:
dir(g)

['DP',
 'DPX',
 'DPY',
 'HP',
 'HPX',
 'HPY',
 'ID',
 'P',
 'P_sat',
 'SP',
 'SPX',
 'SPY',
 'SV',
 'SVX',
 'SVY',
 'T',
 'TD',
 'TDX',
 'TDY',
 'TP',
 'TPX',
 'TPY',
 'T_sat',
 'UV',
 'UVX',
 'UVY',
 'X',
 'Y',
 '__call__',
 '__class__',
 '__copy__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getitem__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__pyx_vtable__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__sizeof__',
 '__slots__',
 '__str__',
 '__subclasshook__',
 '_check_kinetics_species_index',
 '_check_phase_index',
 '_check_reaction_index',
 '_full_states',
 '_init_cti_xml',
 '_init_parts',
 '_references',
 'activities',
 'activity_coefficients',
 'add_reaction',
 'add_species',
 'atomic_weight',
 'atomic_weights',
 'basis',
 'binary_diff_coeffs',
 'chemical_potentials',
 'concentrations',
 'cp',
 'cp_mass',
 'cp_mole',
 'cr

To get help on a specific method, e.g. the `species_index` method:

In [24]:
g.species_index?

For properties, getting the documentation is slightly trickier, as the usual method will give you help for the *result*, e.g.:

In [25]:
g.T?

provides help on Python's `float` class. To get the help for the temperature property, ask for the attribute of the class object itself:

In [26]:
g.__class__.T?

Help can also be obtained using the `help` function:

In [27]:
help(g.species_index)

Help on built-in function species_index:

species_index(...) method of cantera.composite.Solution instance
    ThermoPhase.species_index(self, species) -> int
    
    The index of species *species*, which may be specified as a string or
    an integer. In the latter case, the index is checked for validity and
    returned. If no such species is present, an exception is thrown.



## Chemical Equilibrium

To set a gas mixture to a state of chemical equilibrium, use the `equilibrate` method:

In [28]:
g = ct.Solution('gri30.xml')
g.TPX = 300.0, ct.one_atm, 'CH4:0.95, O2:2, N2:7.52'
g.equilibrate('TP')

The above statement sets the state of object `g` to the state of chemical equilibrium holding temperature and pressure fixed. Alternatively, the specific enthalpy and pressure can be held fixed:

In [29]:
g.TPX = 300.0, ct.one_atm, 'CH4:0.95, O2:2, N2:7.52'
g.equilibrate('HP')

Other options are:
* `'UV'` for fixed specific internal energy and specific volume
* `'SV'` for fixed specific entropy and specific volume
* `'SP'` for fixed specific entropy and pressure

How can you tell if `equilibrate` has correctly found the chemical equilibrium state? One way is to verify that the net rates of progress of all reversible reactions are zero. Here is the code to do this:

In [30]:
g.TPX = 300.0, ct.one_atm, 'CH4:0.95, O2:2, N2:7.52'
g.equilibrate('HP')

In [31]:
rf = g.forward_rates_of_progress
rr = g.reverse_rates_of_progress
for i in range(g.n_reactions):
    if g.is_reversible(i) and rf[i] != 0.0:
        print('  %4i   %10.4g  ' % (i, (rf[i] - rr[i])/rf[i]))

     0   -4.756e-15  
     1   -3.332e-15  
     2   -3.936e-16  
     3   -3.751e-15  
     4   -6.626e-15  
     5   -1.503e-15  
     6    5.377e-15  
     7   -1.609e-15  
     8    1.267e-14  
     9   -1.587e-16  
    10   -4.093e-15  
    11     7.16e-15  
    12   -9.032e-15  
    13    -1.72e-15  
    14   -2.147e-16  
    15   -5.203e-15  
    16   -1.743e-15  
    17    5.791e-15  
    18    1.957e-15  
    19   -5.548e-15  
    20    1.886e-15  
    21    1.421e-15  
    22    1.223e-16  
    23    9.182e-15  
    24    2.055e-15  
    25   -5.119e-15  
    26    3.383e-15  
    27    -1.44e-14  
    28   -3.495e-16  
    29   -1.949e-15  
    30    1.452e-14  
    31    1.048e-14  
    32    3.608e-15  
    33    -3.47e-15  
    34   -3.443e-15  
    35    9.966e-15  
    37    8.611e-16  
    38   -4.981e-15  
    39   -1.486e-15  
    40   -8.614e-15  
    41   -1.518e-15  
    42   -1.232e-15  
    43   -6.832e-15  
    44   -5.306e-15  
    45   -9.065e-15  
    46    

If the magnitudes of the numbers in this list are all very small (which in this case they are), then each reversible reaction is very nearly equilibrated, which only occurs if the gas is in chemical equilibrium.

You might be wondering how `equilibrate` works. (Then again, you might not.) Method `equilibrate` invokes Cantera's chemical equilibrium solver, which uses an element potential method. The element potential method is one of a class of equivalent *nonstoichiometric* methods that all have the characteristic that the probelm reduces to solving a set of $M$ nonlinear algebraic equations, where $M$ is the number of elements (not species). The so-called *stoichiometric* methods, on the other hand (including the Gibbs minimization), require solving $K$ nonlinear equations, where $K$ is the number of species (usually $K >> M$). See Smith and Missen's "Chemical Reaction Equilibrium Analysis" for more information on the various algorithms and their characteristics.

Cantera uses a damped Newton method to solve these equations, and does a few other things to generate a good starting guess and to produce a reasonably robust algorithm. If you want to know more about the details, look at the on-line documentated source code of Cantera C++ class [`ChemEquil`](https://cantera.org/documentation/dev/doxygen/html/da/d0e/classCantera_1_1ChemEquil.html).

## Chemical Kinetics

[`Solution`](https://cantera.org/documentation/dev/sphinx/html/cython/importing.html#cantera.Solution) objects are also [`Kinetics`](https://cantera.org/documentation/dev/sphinx/html/cython/kinetics.html#cantera.Kinetics) objects, and provide all of the methods necessary to compute the thermodynamic quantities associated with each reaction, reaction rates, and species creation and destruction rates. They also provide methods to inspect the quantities that define each reaction, such as the rate constants and the stoichiometric coefficients. The rate calculation functions are used extensively within Cantera's *[reactor network model](https://cantera.org/documentation/dev/sphinx/html/cython/zerodim.html#sec-cython-zerodim)* and *[1D flame model](https://cantera.org/documentation/dev/sphinx/html/cython/onedim.html#sec-cython-onedim)*.

Information about individual reactions that are independent of the thermodynamic state can be obtained by accessing [`Reaction`](https://cantera.org/documentation/dev/sphinx/html/cython/kinetics.html#cantera.Reaction) objects with the [`Kinetics.reaction`](https://cantera.org/documentation/dev/sphinx/html/cython/kinetics.html#cantera.Kinetics.reaction) method:

In [32]:
g = ct.Solution('gri30.cti')
r = g.reaction(2)  # get a Reaction object
r

<ElementaryReaction: H2 + O <=> H + OH>

In [33]:
r.reactants

{'H2': 1.0, 'O': 1.0}

In [34]:
r.products

{'H': 1.0, 'OH': 1.0}

In [35]:
r.rate

Arrhenius(A=38.7, b=2.7, E=2.61918e+07)

If we are interested in only certain types of reactions, we can use this information to filter the full list of reactions to find just the ones of interest. For example, here we find the indices of just those reactions which convert `CO` into `CO2`:

In [36]:
II = [i for i, r in enumerate(g.reactions()) if 'CO' in r.reactants and 'CO2' in r.products]

for i in II:
    print(g.reaction(i).equation)

CO + O (+M) <=> CO2 (+M)
CO + O2 <=> CO2 + O
CO + OH <=> CO2 + H
CO + HO2 <=> CO2 + OH


(Actually, we should also include the reactions where the reaction is written such that `CO2` is a reactant and `CO` is a product, but for this example, we'll just stick to this smaller set of reactions.) Now, let's set the composition to an interesting state:

In [37]:
g.TPX = 300, 101325, {'CH4':0.6, 'O2':1.0, 'N2':3.76}
g.equilibrate('HP')

We can verify that this is an equilibrium state by seeing that the net reactions rates are essentially zero:

In [38]:
g.net_rates_of_progress[II]

array([-8.80914265e-20, -1.03761536e-20, -1.77635684e-15, -4.99749439e-20])

Now let's see what happens if we decrease the temperature of the mixture:

In [39]:
g.TP = g.T-100, None
g.net_rates_of_progress[II]

array([3.18644948e-05, 5.00489577e-08, 1.05964923e-01, 2.89502609e-06])

All of the reaction rates are positive, favoring the formation of `CO2` from `CO`, with the third reaction, `CO + OH <=> CO2 + H` proceeding the fastest. If we look at the enthalpy change associated with each of these reactions:

In [40]:
g.delta_enthalpy[II]

array([-5.33034657e+08, -2.23248515e+07, -8.76650086e+07, -2.49169628e+08])

we see that the change is negative in each case, indicating a net release of thermal energy. The total heat release rate can be computed either from the reaction rates:

In [41]:
np.dot(g.net_rates_of_progress, g.delta_enthalpy)

-58013370.72088288

or from the species production rates:

In [42]:
np.dot(g.net_production_rates, g.partial_molar_enthalpies)

-58013370.72088274

The contribution from just the selected reactions is:

In [43]:
np.dot(g.net_rates_of_progress[II], g.delta_enthalpy[II])

-9307123.262565034

or about 16% of the total heat release rate.